### PR TITLE
{ts} count from flatfile

### PR DESCRIPTION
**Pull request by TomSmithCGAT**
Resurrects  #125 

This introduces the `count_tab` command to count reads per gene using the flatfile output from e.g FeatureCounts/HTSeq. Currently `count_tab` does not support per-cell counting.

### CODE REVIEW

File: umi_tools/count_tab.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    parser = U.OptionParser(version="%prog version: $Id$",
+                            usage=globals()["__doc__"])
+
+    # add common options (-h/--help, ...) and parse command line
+    (options, args) = U.Start(parser, argv=argv, add_group_dedup_options=False)
+
+    if options.random_seed:
+        np.random.seed(options.random_seed)
+
+    if options.stdin != sys.stdin:
```
Comment (IanSudbery) [inline umi_tools/count_tab.py]: This should not be neccessary. Experiment.py should do this for you. We only need this if we are handling a BAM

File: umi_tools/count_tab.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            infile,
+            umi_getter=umi_getter):
+
+        umis = counts.keys()
+
+        nInput += sum(counts.values())
+
+        # set up UMIClusterer functor with methods specific to
+        # specified options.method
+        processor = network.UMIClusterer(options.method)
```
Comment (IanSudbery) [inline umi_tools/count_tab.py]: SHuoldn't need to build a new procesor each time

File: umi_tools/count_tab.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import sys
+import collections
+import re
+
+# required to make iteritems python2 and python3 compatible
+from builtins import dict
+
+from functools import partial
+
+import pysam
```
Comment (IanSudbery) [inline umi_tools/count_tab.py]: Looks as if quite a lot of these imports are not neccesary. I don't see pysam used anywhere. Nor pandas, nor numpy.

File: umi_tools/umi_methods.py [hunk 4]
```diff
@@ -1394,6 +1406,47 @@ def __call__(self, inreads):
                 yield self.reads_dict[p][k], k, "bundle"
 
 
+def get_gene_count_tab(infile,
+                       umi_getter=None):
+
+    ''' Yields the counts per umi for each gene
+
+    ignore_umi: don't include the umi in the dict key
```
Comment (IanSudbery) [inline umi_tools/umi_methods.py]: This and skip_regex do not appear to be arguments to `get_gene_count_tab`

File: umi_tools/count_tab.py (hunk 2 above)
Comment (TomSmithCGAT) [inline umi_tools/count_tab.py]: Yep, that's in the wrong place! Cheers

File: umi_tools/count_tab.py (hunk 1 above)
Comment (TomSmithCGAT) [inline umi_tools/count_tab.py]: Agreed lines 87-88 are unnecessary. Could you explain what the issue is with line 90?

File: umi_tools/count_tab.py (hunk 3 above)
Comment (TomSmithCGAT) [inline umi_tools/count_tab.py]: Yes, lazy copying...

File: umi_tools/umi_methods.py (hunk 4 above)
Comment (TomSmithCGAT) [inline umi_tools/umi_methods.py]: Yes, `ignore_umi` is not a required method. `skip_regex` would be useful but not currently available. I've updated documentation to mark `skip_regex` as "TODO". Thanks!

File: umi_tools/count_tab.py (hunk 1 above)
Comment (IanSudbery) [inline umi_tools/count_tab.py]: If `options.stdin != sys.stdin`, then `options.stdin` should already be an open file handle pointing at the file specified by `-S` on the command line. 

image user runs `umi_tools count_tab -S feature_counts.tsv.gz`
Thus, with your code here on line 89 options.stdin is `<FileHandle: gzip <feature_counts.tsv.gz>>`. On line 90 you discover that it is not `stdin`, and so on line 91 you open the file `feature_counts.tsv.gz` again and set it to `infile`. 
Thus on line 95, `infile` contains what `options.stdin` contained on line 89, irrespective of what happens in 90-94, and the whole block could be replaced with a single line `infile=options.stdin`, or even `options.stdin` passed to `get_gene_count_tab` on line 104.
Bascially what

Comment (TomSmithCGAT) [inline umi_tools/count_tab.py]: Ah yes, of course. I was forgetting how Experiment/Utilities handled the I/O options. Thanks!

**Pull request merged**